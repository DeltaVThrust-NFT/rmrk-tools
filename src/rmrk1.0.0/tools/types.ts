import { Send } from "../classes/send";
import { Buy } from "../classes/buy";
import { List } from "../classes/list";
import { NFT } from "../classes/nft";
import { Emote } from "../classes/emote";
import { ChangeIssuer } from "../classes/changeissuer";

export type Options = {
  ws: string;
  from: string;
  to: string;
  prefixes: string;
  blocks: string;
  json: string;
  folder: string;
  append: string;
  remark: string;
};

export type BlockCalls = {
  block: number;
  calls: BlockCall[];
};

export type BlockCall = {
  call: string;
  value: string;
  caller: string;
  extras?: BlockCall[];
};

export enum OP_TYPES {
  BUY = "BUY",
  LIST = "LIST",
  MINT = "MINT",
  MINTNFT = "MINTNFT",
  SEND = "SEND",
  EMOTE = "EMOTE",
  CONSUME = "CONSUME",
  CHANGEISSUER = "CHANGEISSUER",
}

export type Interaction = Send | Buy | List | Emote | ChangeIssuer;
