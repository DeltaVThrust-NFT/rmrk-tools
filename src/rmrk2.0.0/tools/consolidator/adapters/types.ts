import { NFT } from "../../../classes/nft";
import { Collection } from "../../../classes/collection";
import {
  BaseConsolidated,
  CollectionConsolidated,
  NFTConsolidated,
} from "../consolidator";
import { Base } from "../../../classes/base";
import { AcceptEntityType } from "../../../classes/accept";

export interface IConsolidatorAdapter {
  updateNFTEmote(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateNFTList(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateNftResadd(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateEquip(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateNftAccept(
    nft: NFT,
    consolidatedNFT: NFTConsolidated,
    entity: AcceptEntityType
  ): Promise<any>;
  updateNFTBuy(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateNFTSend(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateNFTBurn(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateNFTMint(nft: NFT): Promise<any>;
  updateSetPriority(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateSetAttribute(nft: NFT, consolidatedNFT: NFTConsolidated): Promise<any>;
  updateCollectionMint(collection: CollectionConsolidated): Promise<any>;
  updateBase(base: Base): Promise<any>;
  updateBaseEquippable(
    base: Base,
    consolidatedBase: BaseConsolidated
  ): Promise<any>;
  updateBaseThemeAdd(
    base: Base,
    consolidatedBase: BaseConsolidated
  ): Promise<any>;
  updateCollectionIssuer(
    collection: Collection,
    consolidatedCollection: CollectionConsolidated
  ): Promise<any>;
  updateBaseIssuer(
    base: Base,
    consolidatedBase: BaseConsolidated
  ): Promise<any>;
  updateNFTChildrenRootOwner(nft: NFT): Promise<any>;
  getNFTById(id: string): Promise<NFTConsolidated | undefined>;
  getCollectionById(id: string): Promise<CollectionConsolidated | undefined>;
  getBaseById(id: string): Promise<BaseConsolidated | undefined>;
  getNFTByIdUnique(id: string): Promise<NFTConsolidated | undefined>;
  getAllNFTs?: () => Promise<NFTConsolidated[]>;
  getAllCollections?: () => Promise<CollectionConsolidated[]>;
  getAllBases?: () => Promise<BaseConsolidated[]>;
}
