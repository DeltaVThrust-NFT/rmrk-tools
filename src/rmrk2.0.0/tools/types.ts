import { Send } from "../classes/send";
import { Buy } from "../classes/buy";
import { List } from "../classes/list";
import { Emote } from "../classes/emote";
import { ChangeIssuer } from "../classes/changeissuer";

export type Options = {
  ws: string;
  from: string;
  to: string;
  prefixes: string;
  blocks: string;
  json: string;
  folder: string;
  append: string;
  remark: string;
};

export type BlockCalls = {
  block: number;
  calls: BlockCall[];
};

export type BlockCall = {
  call: string;
  value: string;
  caller: string;
  extras?: BlockCall[];
};

export type BaseType = "svg";

export type Interaction = Send | Buy | List | Emote | ChangeIssuer;
