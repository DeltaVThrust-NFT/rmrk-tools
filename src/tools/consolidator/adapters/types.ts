import { NFT } from "../../../rmrk1.0.0/classes/nft";
import { Collection } from "../../../rmrk1.0.0/classes/collection";
import { CollectionConsolidated, NFTConsolidated } from "../consolidator";

export interface IConsolidatorAdapter {
  updateNFTEmote(
    nft: NFT,
    consolidatedNFT: NFTConsolidated,
    updatedAtBlock: number
  ): Promise<any>;
  updateNFTList(
    nft: NFT,
    consolidatedNFT: NFTConsolidated,
    updatedAtBlock: number
  ): Promise<any>;
  updateNFTBuy(
    nft: NFT,
    consolidatedNFT: NFTConsolidated,
    updatedAtBlock: number
  ): Promise<any>;
  updateNFTSend(
    nft: NFT,
    consolidatedNFT: NFTConsolidated,
    updatedAtBlock: number
  ): Promise<any>;
  updateNFTConsume(
    nft: NFT,
    consolidatedNFT: NFTConsolidated,
    updatedAtBlock: number
  ): Promise<any>;
  updateNFTMint(nft: NFT, updatedAtBlock: number): Promise<any>;
  updateCollectionMint(collection: CollectionConsolidated): Promise<any>;
  updateCollectionIssuer(
    collection: Collection,
    consolidatedCollection: CollectionConsolidated,
    updatedAtBlock: number
  ): Promise<any>;
  getNFTById(id: string): Promise<NFTConsolidated | undefined>;
  getCollectionById(id: string): Promise<CollectionConsolidated | undefined>;
  getNFTByIdUnique(id: string): Promise<NFTConsolidated | undefined>;
  getAllNFTs?: () => Promise<Record<string, NFTConsolidated>>;
  getAllCollections?: () => Promise<Record<string, CollectionConsolidated>>;
}
