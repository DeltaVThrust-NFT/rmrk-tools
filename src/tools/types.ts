export type Options = {
  ws: string;
  from: string;
  to: string;
  prefixes: string;
  blocks: string;
  json: string;
  folder: string;
  append: string;
};

export type BlockCalls = {
  block: number;
  calls: BlockCall[];
};

export type BlockCall = {
  call: string;
  value: string;
  caller: string;
};
