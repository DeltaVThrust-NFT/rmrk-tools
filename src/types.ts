export type DisplayType =
  | "boost_number"
  | "boost_percentage"
  | "number"
  | "date";

export interface Attribute {
  display_type?: DisplayType;
  trait_type?: string;
  value: number | string;
  max_value?: number;
  mutable?: boolean;
}
