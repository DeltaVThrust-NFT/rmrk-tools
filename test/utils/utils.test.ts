import { blocks489x_630x } from "../mocks/blocks-dump";
import {
  getRemarksFromBlocks,
  stringIsAValidUrl,
  prefixToArray,
  isSystemRemark,
  isUtilityBatch,
  getRemarkData,
} from "../../src/tools/utils";
import { Call as TCall } from "@polkadot/types/interfaces";

// Test getRemarksFromBlocks
describe("utils: getRemarksFromBlocks", () => {
  it("should return remarks from blocks", () => {
    expect(getRemarksFromBlocks(blocks489x_630x)).toMatchSnapshot();
  });
});

describe("utils: stringIsAValidUrl", () => {
  it("should check if string is a URL and return true", () => {
    expect(stringIsAValidUrl("https://rmrk.app/")).toBeTruthy();
  });
});

describe("utils: stringIsAValidUrl", () => {
  it("should check if string is a URL and return false", () => {
    expect(stringIsAValidUrl("wrong url")).toBeFalsy();
  });
});

describe("utils: prefixToArray", () => {
  it("should take prefix string and split it into an array", () => {
    expect(prefixToArray("0x726d726b,0x524d524b")).toEqual([
      "0x726d726b",
      "0x524d524b",
    ]);
  });
});

describe("utils: prefixToArray", () => {
  it("should take prefix string and split it into an array and convert string to hex", () => {
    expect(prefixToArray("0x726d726b,RMRK")).toEqual([
      "0x726d726b",
      "0x524d524b",
    ]);
  });
});

describe("utils: isSystemRemark", () => {
  it("should check if passed data is a remark and return a boolean - true", () => {
    expect(
      isSystemRemark(
        {
          args: ["0x726d726b", "0x524d524b"],
          section: "system",
          method: "remark",
        } as any,
        ["0x726d726b", "0x524d524b"]
      )
    ).toBeTruthy();
  });
});

describe("utils: isSystemRemark", () => {
  it("should check if passed data is a remark and return a boolean - false", () => {
    expect(
      isSystemRemark(
        {
          args: "",
          section: "system",
          method: "",
        } as any,
        ["0x726d726b", "0x524d524b"]
      )
    ).toBeFalsy();
  });
});

describe("utils: isUtilityBatch", () => {
  it("should check if passed a utility and return a boolean - true", () => {
    expect(
      isUtilityBatch({
        section: "utility",
        method: "batch",
      } as TCall)
    ).toBeTruthy();
  });
});

describe("utils: isUtilityBatch", () => {
  it("should check if passed a utility and return a boolean - true", () => {
    expect(
      isUtilityBatch({
        section: "utility",
        method: "batchAll",
      } as TCall)
    ).toBeTruthy();
  });
});

describe("utils: isUtilityBatch", () => {
  it("should check if passed a utility and return a boolean - false", () => {
    expect(
      isUtilityBatch({
        section: "remark",
        method: "batchAll",
      } as TCall)
    ).toBeFalsy();
  });
});

describe("utils: isUtilityBatch", () => {
  it("should check if passed a utility and return a boolean - false", () => {
    expect(
      isUtilityBatch({
        section: "utility",
        method: "wrong",
      } as TCall)
    ).toBeFalsy();
  });
});

describe("utils: getRemarkData", () => {
  it("should take a remark string and turn into json", () => {
    const remark =
      '{"collection"%3A"241B8516516F381A-OKSM"%2C"name"%3A"Kusama Octahedron"%2C"transferable"%3A1%2C"sn"%3A"0000000000000004"%2C"metadata"%3A"ipfs%3A//ipfs/QmXwp5VsPmTdWvFKmc9VwnFkp9jN6ktFiKc5tSMHCuN4pW"}';
    expect(getRemarkData(remark)).toMatchSnapshot();
  });
});
