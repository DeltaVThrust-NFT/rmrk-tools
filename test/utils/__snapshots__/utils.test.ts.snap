// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils: getRemarkData should take a remark string and turn into json 1`] = `
Object {
  "collection": "241B8516516F381A-OKSM",
  "metadata": "ipfs://ipfs/QmXwp5VsPmTdWvFKmc9VwnFkp9jN6ktFiKc5tSMHCuN4pW",
  "name": "Kusama Octahedron",
  "sn": "0000000000000004",
  "transferable": 1,
}
`;

exports[`utils: getRemarksFromBlocks should return remarks from blocks 1`] = `
Array [
  Object {
    "block": 4892957,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "MINT",
    "remark": "rmrk::MINT::{\\"version\\"%3A\\"RMRK0.1\\"%2C\\"name\\"%3A\\"Genesis limited edition obxium art NFTs on Kusama blockchain\\"%2C\\"max\\"%3A5%2C\\"issuer\\"%3A\\"DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby\\"%2C\\"symbol\\"%3A\\"OKSM\\"%2C\\"id\\"%3A\\"241B8516516F381A-OKSM\\"%2C\\"metadata\\"%3A\\"ipfs%3A//ipfs/QmTcaAPWPY5NinmCdDJAi6YFmyag41UEy4SpE1jn4Xdhnx\\"}",
    "version": "0.1",
  },
  Object {
    "block": 4892977,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "MINTNFT",
    "remark": "rmrk::MINTNFT::{\\"collection\\"%3A\\"241B8516516F381A-OKSM\\"%2C\\"name\\"%3A\\"Kusama Cube\\"%2C\\"transferable\\"%3A1%2C\\"sn\\"%3A\\"0000000000000001\\"%2C\\"metadata\\"%3A\\"ipfs%3A//ipfs/QmQGFnoHCRtAcPaWcoiTHHTJ7TEhrpur5WbZSjZdeY3bEp\\"}",
    "version": "0.1",
  },
  Object {
    "block": 4893005,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "MINTNFT",
    "remark": "rmrk::MINTNFT::{\\"collection\\"%3A\\"241B8516516F381A-OKSM\\"%2C\\"name\\"%3A\\"Kusama Dodecahedron\\"%2C\\"transferable\\"%3A1%2C\\"sn\\"%3A\\"0000000000000003\\"%2C\\"metadata\\"%3A\\"ipfs%3A//ipfs/QmUn5sn7rZwzoT65KZZ3pfmE9SoeK556R1z7715yqYWvHW\\"}",
    "version": "0.1",
  },
  Object {
    "block": 4893017,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "MINTNFT",
    "remark": "rmrk::MINTNFT::{\\"collection\\"%3A\\"241B8516516F381A-OKSM\\"%2C\\"name\\"%3A\\"Kusama Octahedron\\"%2C\\"transferable\\"%3A1%2C\\"sn\\"%3A\\"0000000000000004\\"%2C\\"metadata\\"%3A\\"ipfs%3A//ipfs/QmXwp5VsPmTdWvFKmc9VwnFkp9jN6ktFiKc5tSMHCuN4pW\\"}",
    "version": "0.1",
  },
  Object {
    "block": 4893031,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "MINTNFT",
    "remark": "rmrk::MINTNFT::{\\"collection\\"%3A\\"241B8516516F381A-OKSM\\"%2C\\"name\\"%3A\\"Kusama Icosahedron\\"%2C\\"transferable\\"%3A1%2C\\"sn\\"%3A\\"0000000000000005\\"%2C\\"metadata\\"%3A\\"ipfs%3A//ipfs/QmS4mDbM1nT2bdqr3su8GWtJSRxWTdtA5NGjg55xjXqdRX\\"}",
    "version": "0.1",
  },
  Object {
    "block": 4893349,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "SEND",
    "remark": "rmrk::SEND::0.1::241B8516516F381A-OKSM-Kusama%20Cube-0000000000000001%3A%3ACpjsLDC1JFyrhm3ftC9Gs4QoyrkHKhZKtK7YqGTRFtTafgp",
    "version": "0.1",
  },
  Object {
    "block": 4960562,
    "caller": "DmUVjSi8id22vcH26btyVsVq39p8EVPiepdBEYhzoLL8Qby",
    "extra_ex": undefined,
    "interaction_type": "SEND",
    "remark": "rmrk::SEND::0.1::241B8516516F381A-OKSM-Kusama%20Tetrahedron-0000000000000002%3A%3AFac4Kv1dQYWTSt4TqU2SiKuLXs8Sj4qwsXNqdDAvYicmiu9",
    "version": "0.1",
  },
  Object {
    "block": 6302530,
    "caller": "DUyAjx5KwFrv7JbdUtvTzvHJFsTaEzzgyeVTUp9TjatS8Lx",
    "extra_ex": undefined,
    "interaction_type": "MINTNFT",
    "remark": "RMRK::MINTNFT::RMRK1.0.0::{\\"collection\\"%3A\\"282781680602E07B32-FUT\\"%2C\\"sn\\"%3A\\"0000000000000001\\"%2C\\"transferable\\"%3A1%2C\\"name\\"%3A\\"Future %231\\"%2C\\"metadata\\"%3A\\"ipfs%3A%2F%2Fipfs%2FQmPgW6zX5Depv5QRmLXzGGVQ32XYrtJpFPmkUVxxhb5NA2\\"%2C\\"currentOwner\\"%3A\\"DUyAjx5KwFrv7JbdUtvTzvHJFsTaEzzgyeVTUp9TjatS8Lx\\"%2C\\"instance\\"%3A\\"FUTURE_1\\"}",
    "version": "0.1",
  },
  Object {
    "block": 6302581,
    "caller": "DUyAjx5KwFrv7JbdUtvTzvHJFsTaEzzgyeVTUp9TjatS8Lx",
    "extra_ex": undefined,
    "interaction_type": "MINTNFT",
    "remark": "RMRK::MINTNFT::RMRK1.0.0::{\\"collection\\"%3A\\"282781680602E07B32-FUT\\"%2C\\"sn\\"%3A\\"0000000000000003\\"%2C\\"transferable\\"%3A1%2C\\"name\\"%3A\\"Blade runner\\"%2C\\"metadata\\"%3A\\"ipfs%3A%2F%2Fipfs%2FQmUg7a7juDxc7hfk7cuY5scx7StXbXHJfHEHB3FUQxJBKJ\\"%2C\\"currentOwner\\"%3A\\"DUyAjx5KwFrv7JbdUtvTzvHJFsTaEzzgyeVTUp9TjatS8Lx\\"%2C\\"instance\\"%3A\\"BLADE_RUNNER\\"}",
    "version": "0.1",
  },
  Object {
    "block": 6302622,
    "caller": "DUyAjx5KwFrv7JbdUtvTzvHJFsTaEzzgyeVTUp9TjatS8Lx",
    "extra_ex": undefined,
    "interaction_type": "MINT",
    "remark": "RMRK::MINT::RMRK1.0.0::{\\"name\\"%3A\\"Future%232\\"%2C\\"max\\"%3A3%2C\\"symbol\\"%3A\\"FUT\\"%2C\\"version\\"%3A\\"RMRK1.0.0\\"%2C\\"issuer\\"%3A\\"DUyAjx5KwFrv7JbdUtvTzvHJFsTaEzzgyeVTUp9TjatS8Lx\\"%2C\\"metadata\\"%3A\\"ipfs%3A%2F%2Fipfs%2FQmNMgKU9GnSam44k8xzif21f91S2GSWR5ATZbPQFsji8aD\\"%2C\\"id\\"%3A\\"282781680602E07B32-FUT\\"}",
    "version": "0.1",
  },
]
`;
